sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,210 @@
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(CancellationToken cancellationToken) =>
+            await WaitForCompletionAsync(DefaultPollingInterval, cancellationToken).ConfigureAwait(false);
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                var serverDelay = response.Headers.TryGetValue(RetryAfterHeaderName, out string retryAfterValue)
+                    && int.TryParse(retryAfterValue, out int serverDelayInSeconds)
+                    ? TimeSpan.FromSeconds(serverDelayInSeconds) : TimeSpan.Zero;
+
+                var delay = serverDelay > pollingInterval
+                    ? serverDelay : pollingInterval;
+
+                await WaitAsync(delay, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        protected virtual async Task WaitAsync(TimeSpan delay, CancellationToken cancellationToken) =>
+            await Task.Delay(delay, cancellationToken);
+
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = ClientDiagnostics.CreateScope($"{OperationTypeName}.UpdateStatus");
+
+            foreach (KeyValuePair<string, string> attribute in ScopeAttributes)
+            {
+                scope.AddAttribute(attribute.Key, attribute.Value);
+            }
+
+            scope.Start();
+
+            try
+            {
+                return await UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+            }
+            catch (RequestFailedException e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+            catch (Exception e)
+            {
+                var requestFailedException = new RequestFailedException("Unexpected failure.", e);
+                scope.Failed(requestFailedException);
+                throw requestFailedException;
+            }
+        }
+
+        private async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+        {
+            var state = await _operation.UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+
+            lock (SetStateLock)
+            {
+                RawResponse = state.RawResponse;
+
+                if (state.Succeeded == true)
+                {
+                    Value = state.Value;
+                    HasCompleted = true;
+                }
+                else if (state.Succeeded == false)
+                {
+                    OperationFailedException = state.OperationFailedException;
```

Comment:
We might want to check that exception is not null or create a request failed exception using clientDiagnostics here. The default behavior might be good enough for most services and would include additional info like headers and content.

Reply:
Added
```cs
_diagnostics.CreateRequestFailedException(state.RawResponse); // added both sync and async versions
```
for the default exception.